src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java

```diff
@@ -0,0 +1,976 @@
+                return;
+            }
+        }
+        if (con instanceof Boolean) {
+            emitIconstInsn((boolean) con ? 1 : 0);
+            return;
+        }
+        // fall through:
+        mv.visitLdcInsn(con);
+    }
+
+    private void emitIconstInsn(int cst) {
+        if (cst >= -1 && cst <= 5) {
+            mv.visitInsn(Opcodes.ICONST_0 + cst);
+        } else if (cst >= Byte.MIN_VALUE && cst <= Byte.MAX_VALUE) {
+            mv.visitIntInsn(Opcodes.BIPUSH, cst);
+        } else if (cst >= Short.MIN_VALUE && cst <= Short.MAX_VALUE) {
+            mv.visitIntInsn(Opcodes.SIPUSH, cst);
+        } else {
+            mv.visitLdcInsn(cst);
+        }
+    }
+
+    private void emitLoad(Class<?> type, int index) {
+        emitLoad(BasicType.of(type), index);
+    }
+
+    private void emitLoad(BasicType type, int index) {
+        int opcode = loadOpcode(type);
+        mv.visitVarInsn(opcode, index);
+    }
+
+    private static int loadOpcode(BasicType type) throws InternalError {
+        return switch (type) {
+            case Z, B, S, C, I -> ILOAD;
+            case J -> LLOAD;
+            case F -> FLOAD;
+            case D -> DLOAD;
+            case L -> ALOAD;
+        };
+    }
+
+    private void emitStore(BasicType type, int index) {
+        int opcode = storeOpcode(type);
+        mv.visitVarInsn(opcode, index);
+    }
+
+    private static int storeOpcode(BasicType type) throws InternalError {
+        return switch (type) {
+            case Z, B, S, C, I -> ISTORE;
+            case J -> LSTORE;
+            case F -> FSTORE;
+            case D -> DSTORE;
+            case L -> ASTORE;
+        };
+    }
+
+    private void emitReturn(Class<?> type) {
+        int opcode = returnOpcode(BasicType.of(type));
+        mv.visitInsn(opcode);
+    }
+
+    private static int returnOpcode(BasicType type) {
+       return switch (type) {
+            case Z, B, S, C, I -> IRETURN;
+            case J -> LRETURN;
+            case F -> FRETURN;
+            case D -> DRETURN;
+            case L -> ARETURN;
+        };
+    }
+
+    private enum BasicType {
+        Z, B, S, C, I, J, F, D, L;
+
+        static BasicType of(Class<?> cls) {
+            if (cls == boolean.class) {
+                return Z;
+            } else if (cls == byte.class) {
+                return B;
+            } else if (cls == short.class) {
+                return S;
+            } else if (cls == char.class) {
+                return C;
+            } else if (cls == int.class) {
+                return I;
+            } else if (cls == long.class) {
+                return J;
+            } else if (cls == float.class) {
+                return F;
+            } else if (cls == double.class) {
+                return D;
+            } else {
+                return L;
+            }
+        }
+    }
+
+    // unaligned constants
+    public final static ValueLayout.OfBoolean JAVA_BOOLEAN_UNALIGNED = JAVA_BOOLEAN;
```

Comment:
as far as i understand, those constant are accessed from the bytecode, they should be in their own class to cleanly separate the specializer part and the runtime part.

Reply:
I've put these in a nest `Runtime` class, with a comment explaining that they are referenced from the generated code.